docs/DESIGN.md

```diff
@@ -0,0 +1,532 @@
+
+The server must validate the certification path from a trust anchor to the
+client's certificate during authentication, prior to the authorisation 
+subsystem. Recall that the server has no responsibility to verify these subject
+alternative names, as per
+[rfc5280](https://datatracker.ietf.org/doc/html/rfc5280#section-4.2.1.6)
+that is the job of the CA.
+
+At a minimum, a server must support reading the following kinds of SANs out of
+the client's x509 cert to define `Ids(c)`:
+- email addresses
+- DNS names
+A server may choose to implement support additional kinds of SANs defined in
+[rfc5280](https://datatracker.ietf.org/doc/html/rfc5280#section-4.2.1.6).
+
+Server implementations are encouraged to consider normalisation of SAN data
+when implementing their authorisation subsystems.
+
+A single x509 client certificate for an authenticated client may bind 0, 1 or
+many SANs to the client's public key. The set of upstreams that a client is
+authorised to be forwarded to is defined as the union, over the SANs bound to
+the client's certificate, of the sets of upstreams authorised for each SAN.
+
+In the trivial case where a client certificate binds 0 SANs, then the set of
+upstreams that the client is authorised to be forwarded to will be the empty
+set, i.e., the client will not be authorised to be forwarded to any upstream.
+
+The remaining pieces of input data for authorisation, `Groups(.)`, `UGroups(.)`
+and `Upstreams(.)`, will be sourced by the proof-of-concept server from a 
+JSON configuration file in a path specified by the user.
+
+#### proof-of-concept authorisation data format
+
+Example:
+
+```
+{
+    "meta": {
+        "format": "tcplb_authz",
+        "version": 0
+    },
+    "idtype_clientid_group": [
+        {"t": "common_name", "c": "Alice", "g": "admin"},
+        {"t": "common_name", "c": "Bob", "g": "beta"},
+        {"t": "common_name", "c": "Bob", "g": "alpha"}
+    ],
+    "group_can_forward_to_upstreamgroup": [
+        {"g": "alpha", "ug": "web"},
+        {"g": "beta", "ug": "worker"},
+        {"g": "admin", "ug": "web"},
+        {"g": "admin", "ug": "worker"}
+    ],
+    "upstreamgroup_upstream": [
+        {"ug": "web", "u": "127.0.0.1:1234"},
+        {"ug": "web", "u": "127.0.0.1:4567"},
+        {"ug": "worker", "u": "127.0.0.1:5678"},
+        {"ug": "worker", "u": "127.0.0.1:5679"}
+    ]
+}
+```
+- the "meta" section supports future evolution of this configuration format
+- "idtype_clientid_group" declares which client identity ("c") of
+  which identity type ("t") belongs to which group ("g").
+- "group_can_forward_to_upstreamgroup" declares which group ("g") is allowed to
+  be forwarded to which upstream group ("ug"),
+- "upstreamgroup_upstream" declares which upstream ("u") belongs
+  to which upstream group ("ug").
+
+A [version 0 schema](../schemas/tcplb_authz_v0.json) for this data is proposed.
+A proof of concept server may choose not to implement schema validation.
```

Comment:
don't feel obliged to review the schema, i guess the main point i feel strongly about is that whenever someone invents a new format to inflict upon the world, they must also write a spec for the format, and at a minimum the new spec needs a versioning mechanism to ease change when we realise the spec is wrong/limited